### PR TITLE
feat(datepicker): add datepicker component

### PR DESCRIPTION
**Pull request by rossmoody**
## Description

This PR adds the Datepicker component to the website. 

## Checklist

- [x] :ok_hand: website updates are Garden Designer approved (add the designer as a reviewer)
- [x] :black_nib: copy updates are approved (add the content strategist as a reviewer)
- [x] :link: considered opportunities for adding cross-reference URLs (grep for keywords)
- [x] :wheelchair: analyzed via [axe](https://www.deque.com/axe/) and evaluated using VoiceOver
- [x] :zap: audited via [web.dev](https://web.dev/measure/) for performance and SEO
- [x] :memo: tested in Chrome, Firefox, Safari, Edge, and IE11

### CODE REVIEW

File: src/examples/datepicker/datepicker/DatepickerLocalization.tsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+const Example = () => {
+  const [state, setState] = useState(new Date());
+
+  return (
+    <Row justifyContent="center">
+      <Col sm={6}>
+        <Field>
+          <Label>Select a date</Label>
+          <Hint>Select a date that has been localized to Arabic</Hint>
```
Comment (allisonacs) [inline src/examples/datepicker/datepicker/DatepickerLocalization.tsx:21]: Hint is repetitive. Maybe:

**Select a date**
Arabic localization

File: src/pages/components/datepicker.mdx [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+</Usage>
+
+## How to use it
+
+### Default
+
+The typical usage of a Datepicker component.
+
+import Default from '../../examples/datepicker/datepicker/DatepickerDefault.tsx';
+import DefaultCode from '!!raw-loader!../../examples/datepicker/datepicker/DatepickerDefault.tsx';
```
Comment (allisonacs) [inline src/pages/components/datepicker.mdx:32]: Can we have the datepicker be the same width as the input? It's not quite the same size, which feels unresolved. 
image

**Review by allisonacs — CHANGES_REQUESTED** [review]

File: src/pages/components/datepicker.mdx (hunk 2 above)
Comment (austin94) [reply src/pages/components/datepicker.mdx:32]: I would lean toward making the input larger actually. These shouldn't every align width unless it's a coincidence.

Comment (rossmoody) [reply src/pages/components/datepicker.mdx]: I matched the Text input sizing from forms for now but don't feel strongly about either way.

Comment (allisonacs) [reply src/pages/components/datepicker.mdx]: Gotcha. OK, then leaving as is is fine.

**Review by allisonacs — APPROVED** [review]

File: src/examples/datepicker/DatepickerRange.tsx [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import { Field, Label, Input, Message } from '@zendeskgarden/react-forms';
+import { compareAsc, addDays } from 'date-fns';
+import { Grid, Row, Col } from '@zendeskgarden/react-grid';
+
+const StyledCol = styled(Col)`
+  margin-bottom: ${p => p.theme.space.md};
+  max-width: 300px;
+`;
+
+const Example = () => {
```
Comment (jzempel) [inline src/examples/datepicker/DatepickerRange.tsx:20]: I've tweaked this example with a sandbox that I think provides an improved layout across devices along with support for `isCompact` styling: https://codesandbox.io/s/suspicious-johnson-34wks?file=/src/Example.tsx
...and more closely aligns with recommended grid [nesting](https://garden.zendesk.com/components/grid#nesting).

If we don't want the fields to remain at constant width with associated calendars, let's work on improving the expected responsive behavior (i.e. stacked fields).

File: src/examples/datepicker/DatepickerRange.tsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+ *
+ * Use of this source code is governed under the Apache License, Version 2.0
+ * found at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+import React, { useState } from 'react';
+import styled from 'styled-components';
+import { DatepickerRange } from '@zendeskgarden/react-datepickers';
+import { Field, Label, Input, Message } from '@zendeskgarden/react-forms';
+import { compareAsc, addDays } from 'date-fns';
```
Comment (jzempel) [inline src/examples/datepicker/DatepickerRange.tsx:12]: Should `date-fns` be added to the codesandbox parameters?

File: src/examples/datepicker/DatepickerSize.tsx [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+          <Datepicker value={state} onChange={setState}>
+            <Input />
+          </Datepicker>
+        </Field>
+      </Col>
+      <Col sm={5}>
+        <Field>
+          <Label>Compact</Label>
+          <Datepicker value={state} onChange={setState} isCompact>
+            <Input />
```
Comment (jzempel) [inline src/examples/datepicker/DatepickerSize.tsx:30]: Associated `Input` should also be marked `isCompact`.

File: src/examples/datepicker/DatepickerSize.tsx [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    <Row justifyContent="center">
+      <Col sm={5}>
+        <Field>
+          <Label>Default</Label>
+          <Datepicker value={state} onChange={setState}>
+            <Input />
+          </Datepicker>
+        </Field>
+      </Col>
+      <Col sm={5}>
```
Comment (jzempel) [inline src/examples/datepicker/DatepickerSize.tsx:26]: Spacing is needed at mobile sizes. See https://garden.zendesk.com/components/input#size for details.

File: src/pages/components/datepicker.mdx [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+## API
+
+### Datepicker
+
+<PropSheet components={props.data.mdx.components} componentName="datepicker" />
+
+### DatepickerRange
+
+<PropSheet components={props.data.mdx.components} componentName="datepickerrange" />
+
```
Comment (jzempel) [inline src/pages/components/datepicker.mdx:110]: Need H4 placeholders for `DatepickerRange.Calendar`, `DatepickerRange.End`, and `DatepickerRange.Start`

File: src/pages/components/datepicker.mdx [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+<CodeExample code={DefaultCode}>
+  <Default />
+</CodeExample>
+
+### Custom date format
+
+Dates are formatted using the
+[Intl.DateTimeFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DateTimeFormat)
+utility. For other formats, consider libraries like
+[momentjs](https://momentjs.com/) or [date-fns](https://date-fns.org/).
```
Comment (jzempel) [inline src/pages/components/datepicker.mdx:38]: Recommend switching the order of `date-fns` and `momentjs` throughout, to slightly emphasize Garden's preference.

File: src/pages/components/datepicker.mdx [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+</CodeExample>
+
+### Localization
+
+Dates are localized using the
+[Intl.DateTimeFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DateTimeFormat)
+utility. For other formats, use `customParseDate` prop, and formatting libraries
+like [momentjs](https://momentjs.com/) or [date-fns](https://date-fns.org/).
+Customize language, region, and calendar type with a [BCP 47 language
+tag](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DateTimeFormat#Parameters).
```
Comment (jzempel) [inline src/pages/components/datepicker.mdx:66]: This sentence isn't harmonizing nicely with the provided example. "BCP 47 language tag" !== `locale` prop, causing confusion. I'd recommend removing altogether, allowing the example to stand on its own. Someone can learn more about the `locale` prop by visiting the API section below where there is a link to help learn about the "Unicode BCP 47 locale identifier."

**Review by jzempel — CHANGES_REQUESTED** [review]

File: src/examples/datepicker/DatepickerRange.tsx (hunk 3 above)
Comment (rossmoody) [reply src/examples/datepicker/DatepickerRange.tsx:20]: TY 👍

File: src/examples/datepicker/DatepickerRange.tsx (hunk 4 above)
Comment (rossmoody) [reply src/examples/datepicker/DatepickerRange.tsx:12]: It looks like it's a dependency of the [Datepicker package](https://github.com/zendeskgarden/react-components/blob/main/packages/datepickers/package.json#L26). I'm not sure of this answer, @austin94 what say ye?

Comment (austin94) [reply src/examples/datepicker/DatepickerRange.tsx:12]: It's necessary for some of the logic used in the example. I could see value including it for the codesandbox examples.

File: src/pages/components/datepicker.mdx (hunk 8 above)
Comment (austin94) [reply src/pages/components/datepicker.mdx:38]: Also, momentjs just entered "I'm not dead _yet_" mode https://momentjs.com/docs/

File: src/components/MarkdownProvider/components/CodeExample/utils/retrieveCodesandboxParameters.ts [hunk 10]
```diff
@@ -64,6 +64,7 @@ const packageJson = `
     "@zendeskgarden/react-breadcrumbs": "^8.x",
     "@zendeskgarden/react-buttons": "^8.x",
     "@zendeskgarden/react-chrome": "^8.x",
+    "@zendeskgarden/react-datepickers": "^8.x",
```
Comment (jzempel) [inline src/components/MarkdownProvider/components/CodeExample/utils/retrieveCodesandboxParameters.ts:67]: Please also add `date-fns` in here as now it is a direct dependency within some of the examples.

**Review by jzempel — CHANGES_REQUESTED** [review]

File: src/components/MarkdownProvider/components/CodeExample/utils/retrieveCodesandboxParameters.ts (hunk 10 above)
Comment (rossmoody) [reply src/components/MarkdownProvider/components/CodeExample/utils/retrieveCodesandboxParameters.ts:67]: Should I add it to the package.json as well?

Comment (jzempel) [reply src/components/MarkdownProvider/components/CodeExample/utils/retrieveCodesandboxParameters.ts:67]: Nope.

Comment (jzempel) [reply src/components/MarkdownProvider/components/CodeExample/utils/retrieveCodesandboxParameters.ts:67]: Yes, probably should.

**Review by jzempel — APPROVED** [review]

**Pull request merged**